src/coreclr/jit/valuenum.cpp

```diff
@@ -8585,9 +8596,37 @@ void Compiler::fgValueNumberHWIntrinsic(GenTree* tree)
         fgMutateGcHeap(tree DEBUGARG("HWIntrinsic - MemoryStore"));
     }
 
-    int    lookupNumArgs    = HWIntrinsicInfo::lookupNumArgs(hwIntrinsicNode->gtHWIntrinsicId);
-    bool   encodeResultType = vnEncodesResultTypeForHWIntrinsic(hwIntrinsicNode->gtHWIntrinsicId);
-    VNFunc func             = GetVNFuncForNode(tree);
+    VNFunc func         = GetVNFuncForNode(tree);
+    bool   isMemoryLoad = hwIntrinsicNode->OperIsMemoryLoad();
+
+    // If we have a MemoryLoad operation we will use the fgValueNumberByrefExposedLoad
+    // method to assign a value number that depends upon fgCurMemoryVN[ByrefExposed] ValueNumber
+    //
+    if (isMemoryLoad)
```

Comment:
This is for loads which have up to 2 arguments on x86/x64 but up to 3 arguments on ARM64: https://github.com/dotnet/runtime/blob/master/src/coreclr/jit/hwintrinsiclistarm64.h#L171

Is any special handling needed for the GT_LIST? Likewise would the same be true for memory stores which can have 3+ on both x86/x64 and ARM64?

Reply:
I will move up the check for GT_LIST to come before this.
We pessimize the ValueNumber when we have a  GT_LIST case.